modules/core/src/main/java/org/apache/ignite/internal/processors/cache/KeyCacheObjectImpl.java

```diff
@@ -198,7 +217,7 @@ public KeyCacheObjectImpl(Object val, byte[] valBytes, int part) {
 
         KeyCacheObjectImpl other = (KeyCacheObjectImpl)obj;
 
-        return val.equals(other.val);
```

Comment:
Equality of hashcodes is not enough for equality of objects or arrays.
Please,  use Arrays.equals here